### PR TITLE
api/config: use resource type annotations to provide type-to-endpoint.

### LINKED ISSUE #9454: Annotation based mapping between xDS message types and endpoints
Instead of having to hardcode https://github.com/envoyproxy/envoy/blob/master/source/common/config/type_to_endpoint.cc, we should be able to express this relationship via protobuf option annotations, allowing this mapping to be performed programatically.

CC @lizan

### PR DESCRIPTION
**Pull request by htuch**
Previously, type_to_endpoint.cc had a lot of hardcoding, which doesn't scale well with multiple API
versions. See https://github.com/envoyproxy/envoy/pull/9526 for an example of the issues
encountered.

This patch switches to using explicit resource type annotations on service descriptors, which is
great for documentation (previously this was sometimes given in comments, sometimes not), and allows
for a reflection driven reverse map from resource type URL to endpoints to be built at runtime.

Risk level: Low
Testing: New unit tests for type_to_endpoint.cc and golden protoxform tests for the new annotations.

Fixes #9454.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/config/type_to_endpoint.cc [hunk 1]
```diff
@@ -1,94 +1,95 @@
...
+           "envoy.service.secret.v3alpha.SecretDiscoveryService",
+           "envoy.api.v2.ClusterDiscoveryService",
+           "envoy.service.cluster.v3alpha.ClusterDiscoveryService",
+           "envoy.api.v2.EndpointDiscoveryService",
+           "envoy.service.endpoint.v3alpha.EndpointDiscoveryService",
+           "envoy.api.v2.ListenerDiscoveryService",
+           "envoy.service.listener.v3alpha.ListenerDiscoveryService",
+           "envoy.service.discovery.v2.RuntimeDiscoveryService",
+           "envoy.service.runtime.v3alpha.RuntimeDiscoveryService",
+       }) {
```
[the remark this answers is not in the document]
Comment (htuch) [reply source/common/config/type_to_endpoint.cc]: I originally wanted to do this, but can't see a way to do it with `DescriptorPool`, the interface seems relatively abstract and only provides `find` style operations.

File: source/common/config/type_to_endpoint.cc [hunk 2]
```diff
@@ -1,94 +1,95 @@
...
+        service_desc->options().GetExtension(envoy::annotations::resource).type());
+    Service& service = (*type_url_to_service_map)[resource_type_url];
+    for (int method_index = 0; method_index < service_desc->method_count(); ++method_index) {
+      const auto& method_desc = *service_desc->method(method_index);
+      if (absl::StartsWith(method_desc.name(), "Stream")) {
+        service.sotw_grpc_method_ = method_desc.full_name();
+      } else if (absl::StartsWith(method_desc.name(), "Delta")) {
+        service.delta_grpc_method_ = method_desc.full_name();
+      } else if (absl::StartsWith(method_desc.name(), "Fetch")) {
+        service.rest_method_ = method_desc.full_name();
```
[the remark this answers is not in the document]
Comment (htuch) [reply source/common/config/type_to_endpoint.cc]: Sure, will add this.

File: source/common/config/type_to_endpoint.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (htuch) [reply source/common/config/type_to_endpoint.cc]: FYI, here is the interface https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.descriptor#DescriptorPool and some related discussion https://www.mail-archive.com/protobuf@googlegroups.com/msg04540.html

File: source/common/config/type_to_endpoint.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (htuch) [reply source/common/config/type_to_endpoint.cc]: FYI VHDS only does delta xDS today, see https://github.com/envoyproxy/envoy/blob/a60f6853a2c2ebbbfed79dfff0b5b644fd735980/api/envoy/api/v2/rds.proto#L51

File: source/common/config/type_to_endpoint.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/config/type_to_endpoint.cc]: Maybe just add a TODO/comment trail/open a protobuf issue on this?

[the remark this answers is not in the document]
Comment (htuch) [reply source/common/config/type_to_endpoint.cc]: I'l leave a comment, but I think it's almost by design, given that DescriptorPools can be lazily populated and in theory infinitely deep.

**Comment by htuch** [comment]
@mattklein123 https://github.com/envoyproxy/envoy/issues/3091 tracks the service method doc generation, I left a comment to make use of the resource type annotations.

**Pull request merged**